Run HttpServerHandlerTest.assertShardingSphereJDBCTargetChannelReadStartSuccess throw exception

[WARN ] 11:05:32.916 [main] com.zaxxer.hikari.HikariConfig - HikariPool-1 - idleTimeout has been set but has no effect because the pool is operating as a fixed size pool.
[INFO ] 11:05:33.354 [main] ShardingSphere-metadata - Loading 0 tables' meta data.
[ERROR] 11:05:33.503 [ShardingSphere-ShardingScaling-execute-1] o.a.s.s.c.s.ScalingTaskScheduler - Inventory task execute failed.
org.apache.shardingsphere.scaling.core.exception.SyncTaskExecuteException: get estimated rows error.
	at org.apache.shardingsphere.scaling.core.job.task.inventory.InventoryDataScalingTask.getEstimatedRows(InventoryDataScalingTask.java:118)
	at org.apache.shardingsphere.scaling.core.job.task.inventory.InventoryDataScalingTask.start(InventoryDataScalingTask.java:87)
	at org.apache.shardingsphere.scaling.core.job.task.inventory.InventoryDataScalingTaskGroup.start(InventoryDataScalingTaskGroup.java:46)
	at org.apache.shardingsphere.scaling.core.execute.executor.AbstractShardingScalingExecutor.run(AbstractShardingScalingExecutor.java:53)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:69)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.h2.jdbc.JdbcSQLException: Table "t1" not found; SQL statement:
SELECT COUNT(*) FROM t1  [42102-196]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:345)
	at org.h2.message.DbException.get(DbException.java:179)
	at org.h2.message.DbException.get(DbException.java:155)
	at org.h2.command.Parser.readTableOrView(Parser.java:5552)
	at org.h2.command.Parser.readTableFilter(Parser.java:1266)
	at org.h2.command.Parser.parseSelectSimpleFromPart(Parser.java:1946)
	at org.h2.command.Parser.parseSelectSimple(Parser.java:2095)
	at org.h2.command.Parser.parseSelectSub(Parser.java:1940)
	at org.h2.command.Parser.parseSelectUnion(Parser.java:1755)
	at org.h2.command.Parser.parseSelect(Parser.java:1743)
	at org.h2.command.Parser.parsePrepared(Parser.java:449)
	at org.h2.command.Parser.parse(Parser.java:321)
	at org.h2.command.Parser.parse(Parser.java:293)
	at org.h2.command.Parser.prepareCommand(Parser.java:258)
	at org.h2.engine.Session.prepareLocal(Session.java:578)
	at org.h2.engine.Session.prepareCommand(Session.java:519)
	at org.h2.jdbc.JdbcConnection.prepareCommand(JdbcConnection.java:1204)
	at org.h2.jdbc.JdbcPreparedStatement.<init>(JdbcPreparedStatement.java:73)
	at org.h2.jdbc.JdbcConnection.prepareStatement(JdbcConnection.java:288)
	at com.zaxxer.hikari.pool.ProxyConnection.prepareStatement(ProxyConnection.java:316)
	at com.zaxxer.hikari.pool.HikariProxyConnection.prepareStatement(HikariProxyConnection.java)
	at org.apache.shardingsphere.scaling.core.job.task.inventory.InventoryDataScalingTask.getEstimatedRows(InventoryDataScalingTask.java:111)
	... 10 common frames omitted

--- a/shardingsphere-scaling/shardingsphere-scaling-bootstrap/src/test/java/org/apache/shardingsphere/scaling/fixture/FixtureH2JDBCDumper.java
+++ b/shardingsphere-scaling/shardingsphere-scaling-bootstrap/src/test/java/org/apache/shardingsphere/scaling/fixture/FixtureH2JDBCDumper.java
@@ -1,0 +1,38 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.shardingsphere.scaling.fixture;
+
+import org.apache.shardingsphere.scaling.core.config.InventoryDumperConfiguration;
+import org.apache.shardingsphere.scaling.core.datasource.DataSourceManager;
+import org.apache.shardingsphere.scaling.core.execute.executor.dumper.AbstractJDBCDumper;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+
+public final class FixtureH2JDBCDumper extends AbstractJDBCDumper {
+    
+    public FixtureH2JDBCDumper(final InventoryDumperConfiguration dumperConfig, final DataSourceManager dataSourceManager) {
+        super(dumperConfig, dataSourceManager);
+    }
+    
+    @Override
+    protected PreparedStatement createPreparedStatement(final Connection conn, final String sql) throws SQLException {
+        return conn.prepareStatement(sql);
+    }
+}

--- a/shardingsphere-scaling/shardingsphere-scaling-bootstrap/src/test/java/org/apache/shardingsphere/scaling/fixture/FixtureH2ScalingEntry.java
+++ b/shardingsphere-scaling/shardingsphere-scaling-bootstrap/src/test/java/org/apache/shardingsphere/scaling/fixture/FixtureH2ScalingEntry.java
@@ -30,7 +30,7 @@ public final class FixtureH2ScalingEntry implements ScalingEntry {
     
     @Override
     public Class<? extends JDBCDumper> getJdbcDumperClass() {
-        return null;
+        return FixtureH2JDBCDumper.class;
     }
     
     @Override
@@ -45,7 +45,7 @@ public final class FixtureH2ScalingEntry implements ScalingEntry {
     
     @Override
     public Class<? extends Importer> getImporterClass() {
-        return null;
+        return FixtureNopImporter.class;
     }
     
     @Override

--- a/shardingsphere-scaling/shardingsphere-scaling-bootstrap/src/test/java/org/apache/shardingsphere/scaling/fixture/FixtureNopImporter.java
+++ b/shardingsphere-scaling/shardingsphere-scaling-bootstrap/src/test/java/org/apache/shardingsphere/scaling/fixture/FixtureNopImporter.java
@@ -1,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.shardingsphere.scaling.fixture;
+
+import org.apache.shardingsphere.scaling.core.config.ImporterConfiguration;
+import org.apache.shardingsphere.scaling.core.datasource.DataSourceManager;
+import org.apache.shardingsphere.scaling.core.execute.executor.channel.Channel;
+import org.apache.shardingsphere.scaling.core.execute.executor.importer.Importer;
+
+public final class FixtureNopImporter implements Importer {
+    
+    public FixtureNopImporter(final ImporterConfiguration importerConfig, final DataSourceManager dataSourceManager) {
+    }
+    
+    @Override
+    public void setChannel(final Channel channel) {
+    }
+    
+    @Override
+    public void write() {
+    }
+    
+    @Override
+    public void start() {
+    }
+    
+    @Override
+    public void stop() {
+    }
+    
+    @Override
+    public void run() {
+        start();
+        write();
+    }
+}

--- a/shardingsphere-scaling/shardingsphere-scaling-bootstrap/src/test/java/org/apache/shardingsphere/scaling/web/HttpServerHandlerTest.java
+++ b/shardingsphere-scaling/shardingsphere-scaling-bootstrap/src/test/java/org/apache/shardingsphere/scaling/web/HttpServerHandlerTest.java
@@ -27,9 +27,14 @@ import io.netty.handler.codec.http.FullHttpResponse;
 import io.netty.handler.codec.http.HttpMethod;
 import io.netty.handler.codec.http.HttpVersion;
 import io.netty.util.CharsetUtil;
+import lombok.SneakyThrows;
 import org.apache.shardingsphere.scaling.core.config.ScalingConfiguration;
 import org.apache.shardingsphere.scaling.core.config.ScalingContext;
+import org.apache.shardingsphere.scaling.core.config.ScalingDataSourceConfiguration;
 import org.apache.shardingsphere.scaling.core.config.ServerConfiguration;
+import org.apache.shardingsphere.scaling.core.config.SyncConfiguration;
+import org.apache.shardingsphere.scaling.core.datasource.DataSourceManager;
+import org.apache.shardingsphere.scaling.core.utils.SyncConfigurationUtil;
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
@@ -37,8 +42,12 @@ import org.mockito.ArgumentCaptor;
 import org.mockito.Mock;
 import org.mockito.junit.MockitoJUnitRunner;
 
+import javax.sql.DataSource;
 import java.io.InputStream;
 import java.io.InputStreamReader;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.sql.Statement;
 import java.util.HashMap;
 import java.util.Map;
 
@@ -61,11 +70,15 @@ public final class HttpServerHandlerTest {
     
     private ScalingConfiguration scalingConfig;
     
+    private SyncConfiguration syncConfiguration;
+    
     @Before
     public void setUp() {
         initConfig("/config.json");
         ScalingContext.getInstance().init(new ServerConfiguration());
         httpServerHandler = new HttpServerHandler();
+        initTableData(syncConfiguration.getDumperConfiguration().getDataSourceConfiguration());
+        initTableData(syncConfiguration.getImporterConfiguration().getDataSourceConfiguration());
     }
     
     @Test
@@ -168,5 +181,18 @@ public final class HttpServerHandlerTest {
         InputStream fileInputStream = HttpServerHandlerTest.class.getResourceAsStream(configFile);
         InputStreamReader inputStreamReader = new InputStreamReader(fileInputStream);
         scalingConfig = GSON.fromJson(inputStreamReader, ScalingConfiguration.class);
+        syncConfiguration = SyncConfigurationUtil.toSyncConfigurations(scalingConfig).iterator().next();
+    }
+    
+    @SneakyThrows(SQLException.class)
+    private void initTableData(final ScalingDataSourceConfiguration dataSourceConfig) {
+        DataSource dataSource = new DataSourceManager().getDataSource(dataSourceConfig);
+        try (Connection connection = dataSource.getConnection();
+             Statement statement = connection.createStatement()) {
+            statement.execute("DROP TABLE IF EXISTS t1");
+            statement.execute("CREATE TABLE t1 (id INT PRIMARY KEY, user_id VARCHAR(12))");
+            statement.execute("DROP TABLE IF EXISTS t2");
+            statement.execute("CREATE TABLE t2 (id INT PRIMARY KEY, user_id VARCHAR(12))");
+        }
     }
 }

--- a/shardingsphere-scaling/shardingsphere-scaling-bootstrap/src/test/resources/config.json
+++ b/shardingsphere-scaling/shardingsphere-scaling-bootstrap/src/test/resources/config.json
@@ -20,7 +20,7 @@
     "source": {
       "type": "shardingSphereJdbc",
       "parameter": {
-        "dataSource": "dataSources:\n ds_0:\n  dataSourceClassName: com.zaxxer.hikari.HikariDataSource\n  props:\n    jdbcUrl: jdbc:h2:mem:test_db_2;DB_CLOSE_DELAY=-1;DATABASE_TO_UPPER=false;MODE=MySQL\n    username: root\n    password: '123456'\n    connectionTimeout: 30000\n    idleTimeout: 60000\n    maxLifetime: 1800000\n    maxPoolSize: 50\n    minPoolSize: 1\n    maintenanceIntervalMilliseconds: 30000\n    readOnly: false\n",
+        "dataSource": "dataSources:\n ds_0:\n  dataSourceClassName: com.zaxxer.hikari.HikariDataSource\n  props:\n    jdbcUrl: jdbc:h2:mem:test_db_2;DB_CLOSE_DELAY=-1;DATABASE_TO_UPPER=false;MODE=MySQL\n    username: root\n    password: 'password'\n    connectionTimeout: 30000\n    idleTimeout: 60000\n    maxLifetime: 1800000\n    maxPoolSize: 50\n    minPoolSize: 1\n    maintenanceIntervalMilliseconds: 30000\n    readOnly: false\n",
         "rule": "rules:\n- !SHARDING\n  defaultDatabaseStrategy:\n    standard:\n      shardingAlgorithmName: inline\n      shardingColumn: user_id\n  tables:\n    t1:\n      actualDataNodes: ds_0.t1\n      keyGenerateStrategy:\n        column: order_id\n      logicTable: t1\n      tableStrategy:\n        standard:\n          shardingAlgorithmName: inline\n          shardingColumn: order_id\n    t2:\n      actualDataNodes: ds_0.t2\n      keyGenerateStrategy:\n        column: order_item_id\n      logicTable: t2\n      tableStrategy:\n        standard:\n          shardingAlgorithmName: inline\n          shardingColumn: order_id"
       }
     },

--- a/shardingsphere-scaling/shardingsphere-scaling-bootstrap/src/test/resources/config_sharding_sphere_jdbc_target.json
+++ b/shardingsphere-scaling/shardingsphere-scaling-bootstrap/src/test/resources/config_sharding_sphere_jdbc_target.json
@@ -20,14 +20,14 @@
     "source": {
       "type": "shardingSphereJdbc",
       "parameter": {
-        "dataSource": "dataSources:\n ds_0:\n  dataSourceClassName: com.zaxxer.hikari.HikariDataSource\n  props:\n    jdbcUrl: jdbc:h2:mem:test_db_2;DB_CLOSE_DELAY=-1;DATABASE_TO_UPPER=false;MODE=MySQL\n    username: root\n    password: '123456'\n    connectionTimeout: 30000\n    idleTimeout: 60000\n    maxLifetime: 1800000\n    maxPoolSize: 50\n    minPoolSize: 1\n    maintenanceIntervalMilliseconds: 30000\n    readOnly: false\n",
+        "dataSource": "dataSources:\n ds_0:\n  dataSourceClassName: com.zaxxer.hikari.HikariDataSource\n  props:\n    jdbcUrl: jdbc:h2:mem:test_db_2;DB_CLOSE_DELAY=-1;DATABASE_TO_UPPER=false;MODE=MySQL\n    username: root\n    password: 'password'\n    connectionTimeout: 30000\n    idleTimeout: 60000\n    maxLifetime: 1800000\n    maxPoolSize: 50\n    minPoolSize: 1\n    maintenanceIntervalMilliseconds: 30000\n    readOnly: false\n",
         "rule": "rules:\n- !SHARDING\n  defaultDatabaseStrategy:\n    standard:\n      shardingAlgorithmName: inline\n      shardingColumn: user_id\n  tables:\n    t1:\n      actualDataNodes: ds_0.t1\n      keyGenerateStrategy:\n        column: order_id\n      logicTable: t1\n      tableStrategy:\n        standard:\n          shardingAlgorithmName: inline\n          shardingColumn: order_id\n    t2:\n      actualDataNodes: ds_0.t2\n      keyGenerateStrategy:\n        column: order_item_id\n      logicTable: t2\n      tableStrategy:\n        standard:\n          shardingAlgorithmName: inline\n          shardingColumn: order_id"
       }
     },
     "target": {
       "type": "shardingSphereJdbc",
       "parameter": {
-        "dataSource": "dataSources:\n ds_0:\n  dataSourceClassName: com.zaxxer.hikari.HikariDataSource\n  props:\n    jdbcUrl: jdbc:h2:mem:test_db_2;DB_CLOSE_DELAY=-1;DATABASE_TO_UPPER=false;MODE=MySQL\n    username: root\n    password: '123456'\n    connectionTimeout: 30000\n    idleTimeout: 60000\n    maxLifetime: 1800000\n    maxPoolSize: 50\n    minPoolSize: 1\n    maintenanceIntervalMilliseconds: 30000\n    readOnly: false\n",
+        "dataSource": "dataSources:\n ds_0:\n  dataSourceClassName: com.zaxxer.hikari.HikariDataSource\n  props:\n    jdbcUrl: jdbc:h2:mem:test_db_2;DB_CLOSE_DELAY=-1;DATABASE_TO_UPPER=false;MODE=MySQL\n    username: root\n    password: 'password'\n    connectionTimeout: 30000\n    idleTimeout: 60000\n    maxLifetime: 1800000\n    maxPoolSize: 50\n    minPoolSize: 1\n    maintenanceIntervalMilliseconds: 30000\n    readOnly: false\n",
         "rule": "rules:\n- !SHARDING\n  defaultDatabaseStrategy:\n    standard:\n      shardingAlgorithmName: inline\n      shardingColumn: user_id\n  tables:\n    t1:\n      actualDataNodes: ds_0.t1\n      keyGenerateStrategy:\n        column: order_id\n      logicTable: t1\n      tableStrategy:\n        standard:\n          shardingAlgorithmName: inline\n          shardingColumn: order_id\n    t2:\n      actualDataNodes: ds_0.t2\n      keyGenerateStrategy:\n        column: order_item_id\n      logicTable: t2\n      tableStrategy:\n        standard:\n          shardingAlgorithmName: inline\n          shardingColumn: order_id"
       }
     }

